### PR TITLE
Create flag to enable frame by frame planar surfaces

### PR DESCRIPTION
**Pull request by zthorson**
## Description

This will add the capability to allow for PLANAR surfaces to be constructed on a frame by frame basis, rather than locking in on the first good offset estimate. Additional feedback as to when fallback mode is used for a projection has also been added.

Further discussion and samples can be found at the end of the issue discussion here: https://github.com/laxnpander/OpenREALM/issues/57

## Reason

Currently, the `surface_generation.cpp` node will lock on to the first good ground estimate when running in PLANAR mode.  This was originally done to improve stitching quality in cases where the surface planar adjustments were creating worse results.  There appear to be many cases where allowing a frame by frame adjustment of the planar surface does result in a better final stitch, so an option to selectively enable/disable it will allow either to be used.

## Method / Design

* A new parameter was added to 'surface_generation.yaml` which allows enabling of frame by frame adjustments in 2.5D planar mode.
`compute_all_frames` - When in PLANAR mode, estimate the elevation of every frame using sparse data rather than locking on the first - Default: off

* For better feedback in the logs, a log message is now emitted if fallback mode was used.  This should make it easier to tell when accurate vs inaccurate poses were used to project a frame.

* A minor bug in fallback mode that resulted in the first fallback frame setting the surface generation offset for the entire flight (rather than waiting for VSLAM corrected values) was also corrected.

## Testing
Compiled and run on:
* Ubuntu 20.04 - Desktop
* Custom OpenREALM wrapper (non-ROS)

### CODE REVIEW

**Comment by laxnpander** [comment]
@zthorson: Jeah, I pushed some changes today. Unfortunately my documentation is not as excellent as yours. But I try to improve that for future commits. Hope it’s not too much work.

**Comment by zthorson** [comment]
@laxnpander The compilation issue should addressed, though I'm seeing some odd behavior with the new scale divergence detection.  It appears that when the scale divergence is detected and the frame is set to not be an accurate one, the final mosaicing step faults out when it tries to request too much memory.

To me, it appears that something must be incorrect in the georeferencing when we just say a previously accurate pose is no longer accurate.  If I comment out `frame->setPoseAccurate(false)` in the scale check, the fault goes away.

I'm going to keep looking at it, but one way I've addressed it is to not call `pushToBufferAll(frame)` if the divergence is detected.

The logs I get when divergence is detected after we have referenced images:
```
2021-05-07 14:13:20.405 (  19.307s) [Stage [pose_esti]    pose_estimation.cpp:202   WARN| Detected scale divergence.
2021-05-07 14:13:20.405 (  19.307s) [Stage [pose_esti] worker_thread_base.cpp:81       0| Timing [Total]: 43 ms
2021-05-07 14:13:20.405 (  19.307s) [Stage [pose_esti]    pose_estimation.cpp:486      0| Input frame buffer: 3/5
2021-05-07 14:13:20.405 (  19.307s) [Stage [pose_esti]    pose_estimation.cpp:183      0| Processing frame #417 with timestamp 77399!
2021-05-07 14:13:20.405 (  19.307s) [Stage [pose_esti]    pose_estimation.cpp:267      0| Tracking frame #417 in visual SLAM...!
2021-05-07 14:13:20.406 (  19.308s) [Publisher [pose_]    pose_estimation.cpp:713      0| Publishing pose of frame #416...
2021-05-07 14:13:20.407 (  19.308s) [Publisher [pose_]QuickStitchWrapperIOS.c:144      0| Pose Pose : output/pose/visual
2021-05-07 14:13:20.407 (  19.308s) [Publisher [pose_]    pose_estimation.cpp:684      0| VSLAM tracking lost, using GNSS fallback for frame 416!
2021-05-07 14:13:20.407 (  19.308s) [Publisher [pose_]    pose_estimation.cpp:746      0| Publishing frame #416...
2021-05-07 14:13:20.407 (  19.308s) [Publisher [pose_]QuickStitchWrapperIOS.c:161      0| Pose Point: output/pointcloud
2021-05-07 14:13:20.407 (  19.308s) [Publisher [pose_]QuickStitchWrapperIOS.c:150      0| Pose Frame: output/frame
2021-05-07 14:13:20.407 (  19.308s) [Publisher [pose_]    pose_estimation.cpp:573      0| Georeferenced pose:
2021-05-07 14:13:20.407 (  19.308s) [Publisher [pose_]    pose_estimation.cpp:574      0| GNSS: [ 427015.66, 4958971.53, 60.90   89.9]
2021-05-07 14:13:20.407 (  19.308s) [Publisher [pose_]    pose_estimation.cpp:576      0| Visual: [      0.90,      -1.85, -0.02,   -7.4]
2021-05-07 14:13:20.407 (  19.308s) [Publisher [pose_]    pose_estimation.cpp:578      0| Diff: [ 427014.76, 4958973.38, 60.92,   97.3]
2021-05-07 14:13:20.407 (  19.308s) [Publisher [pose_]QuickStitchWrapperIOS.c:156      0| Pose Image: debug/tracked
2021-05-07 14:13:20.407 (  19.308s) [Stage [surface_g] surface_generation.cpp:51       0| Processing frame #416...
2021-05-07 14:13:20.407 (  19.308s) [Stage [surface_g] surface_generation.cpp:60       0| Surface assumption: PLANAR.
2021-05-07 14:13:20.407 (  19.308s) [Stage [surface_g] surface_generation.cpp:210      0| Sparse cloud was utilized to compute an initial projection plane at elevation = 1.01.
2021-05-07 14:13:20.407 (  19.308s) [Stage [surface_g] surface_generation.cpp:69       0| Timing [Compute DSM]: 0 ms
2021-05-07 14:13:20.407 (  19.308s) [Stage [surface_g] surface_generation.cpp:78       0| Timing [Container Add]: 0 ms
2021-05-07 14:13:20.407 (  19.308s) [Stage [surface_g] surface_generation.cpp:80       0| Publishing frame for next stage...
2021-05-07 14:13:20.407 (  19.309s) [Stage [surface_g]QuickStitchWrapperIOS.c:168      0| Surface Frame: output/frame
2021-05-07 14:13:20.407 (  19.309s) [Stage [surface_g] surface_generation.cpp:85       0| Timing [Publish]: 0 ms
2021-05-07 14:13:20.407 (  19.309s) [Stage [ortho_rec]ortho_rectification.cpp:62       0| Processing frame #416...
2021-05-07 14:13:20.407 (  19.309s) [Stage [surface_g] surface_generation.cpp:90       0| Timing [Saving]: 0 ms
2021-05-07 14:13:20.407 (  19.309s) [Stage [ortho_rec]ortho_rectification.cpp:68       0| Resize quotient rq = (elevation.resolution() / GSD) = 5.00
2021-05-07 14:13:20.407 (  19.309s) [Stage [surface_g] worker_thread_base.cpp:81       0| Timing [Total]: 0 ms
2021-05-07 14:13:20.407 (  19.309s) [Stage [ortho_rec]ortho_rectification.cpp:70       0| Large resizing of elevation map detected. Keep in mind that ortho resolution is now >> spatial resolution
2021-05-07 14:13:20.407 (  19.309s) [Stage [ortho_rec]ortho_rectification.cpp:75       0| Timing [Resizing]: 0 ms
2021-05-07 14:13:20.407 (  19.309s) [Stage [ortho_rec]      rectification.cpp:62       0| Processing rectification:
2021-05-07 14:13:20.407 (  19.309s) [Stage [ortho_rec]      rectification.cpp:63       0| - ROI (1.000000, 0.000000, 1.000000, 1.000000)
2021-05-07 14:13:20.407 (  19.309s) [Stage [ortho_rec]      rectification.cpp:64       0| - Dimensions: 6 x 6
2021-05-07 14:13:20.407 (  19.309s) [Stage [ortho_rec]      rectification.cpp:95       0| Image successfully rectified.
2021-05-07 14:13:20.407 (  19.309s) [Stage [ortho_rec]ortho_rectification.cpp:81       0| Timing [Rectify]: 0 ms
2021-05-07 14:13:20.407 (  19.309s) [Stage [ortho_rec]ortho_rectification.cpp:94       0| Timing [Adding]: 0 ms
2021-05-07 14:13:20.408 (  19.309s) [Stage [ortho_rec]QuickStitchWrapperIOS.c:176      0| Ortho Frame: output/frame
2021-05-07 14:13:20.408 (  19.309s) [Stage [ortho_rec]QuickStitchWrapperIOS.c:182      0| Ortho Image: output/rectified
2021-05-07 14:13:20.408 (  19.309s) [Stage [ortho_rec]ortho_rectification.cpp:99       0| Timing [Publish]: 0 ms
2021-05-07 14:13:20.408 (  19.309s) [Stage [mosaicing]          mosaicing.cpp:100      0| Processing frame #416...
2021-05-07 14:13:20.408 (  19.309s) [Stage [ortho_rec]ortho_rectification.cpp:104      0| Timing [Saving]: 0 ms
2021-05-07 14:13:20.408 (  19.309s) [Stage [mosaicing]          mosaicing.cpp:117      0| Adding new map data to global map...
2021-05-07 14:13:20.408 (  19.309s) [Stage [ortho_rec] worker_thread_base.cpp:81       0| Timing [Total]: 1 ms
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.2.0) ../modules/core/src/alloc.cpp:73: error: (-4:Insufficient memory) Failed to allocate 211841061932724 bytes in function 'OutOfMemoryError'
```

**Comment by zthorson** [comment]
@laxnpander 
The core of the issue appears to be that the Camera inside the frame has already been updated with the visual pose, so when the georeferencer accesses the camera and gets the pose and translation matrices, they are the matrices relative to visual slam.  

So, when the system applies georeferencing to the buffer here:
https://github.com/laxnpander/OpenREALM/blob/6be8d2d6a6571a8a79f06dbd84b4ab03991d7950/modules/realm_stages/src/pose_estimation.cpp#L492-L497

It thinks that the pose data is in the world coordinate frame already, so it skips applying the referencing, leaving it in visual coordinates.

Finally, the mosaicing code gets what it thinks are georeferenced coordinates and procedes to add them to the map. However, their UTM coordinates are set to 0,0 causing a huge expansion in map size, and a resulting crash.

Long story short, it can be fixed by adding a way to reset the camera pose inside the frame when you set the frame to an inaccurate one.  

For now I will comment it out in this pull request, and I can help clean it up later.

**Comment by laxnpander** [comment]
@zthorson: Ah yes, that adds up. I was using it with a flag to drop all frames with planar assumption, so I didn’t notice it.

**Comment by laxnpander** [comment]
Hmm, I'll fix the branch on my side first and then we'll see what he has to say about the PR.

**Comment by laxnpander** [comment]
@zthorson: Looks good now, Travis didn't like the dev build being broken even though your PR fixed it. A bit weird though.
